### ISSUE TITLE
Add delay feature to the local rate limit filter

### ISSUE THREAD

**Issue by nitgoy** [envoyproxy/envoy#14264]
*Title*: *Add delay feature to the local rate limit filter*

*Description*:
>When a connection or request is rate limited by the L4/L7 local rate limit filters, instead of immediately rejecting the connection or request it would be useful to reject after pre-configured delay. That way, a malicious user is not able to retry as fast as possible which provides a better DoS protection for Envoy.

I believe the RLS service called by the rate limit filter can implement a delay on its end but it would be nice to have that functionality in Envoy especially for the local rate limit filter but could be shared by the global rate limit filter as well.

**Comment by mattklein123** [comment]
I think this is a reasonable request, though I would personally rather see this get built into a generic WAF filter that supports multiple actions (tar pit being one): https://github.com/envoyproxy/envoy/issues/7918

**Comment by nitgoy** [comment]
Regarding WAF, what if the action needs to be taken based on the rate of requests? That means you might need to duplicate the logic of the rate limit filter in WAF. Or did you mean the rate limit filter's decision could be hooked to WAF filter which can then take a appropriate action? IMO that could introduce a lot of coupling between the filters but would be worthwhile if a lot of WAF actions are also useful for rate limit filter.

Do you think it's a good idea to add the functionality to the rate limit filter for now which should be simple enough, and later migrate to what you proposed once WAF is ready? Once WAF is ready, I guess you'd also want to migrate the current RL features like adding custom response headers, to WAF.

**Comment by mattklein123** [comment]
> Or did you mean the rate limit filter's decision could be hooked to WAF filter which can then take a appropriate action? IMO that could introduce a lot of coupling between the filters but would be worthwhile if a lot of WAF actions are also useful for rate limit filter.

I think we could use metadata to allow other filters to cause WAF actions and chain them this way.

> Do you think it's a good idea to add the functionality to the rate limit filter for now which should be simple enough, and later migrate to what you proposed once WAF is ready? Once WAF is ready, I guess you'd also want to migrate the current RL features like adding custom response headers, to WAF.

I would rather just build the WAF filter even if initially it just does one thing and we allow actions to be triggered by metadata from other filters. But I think it would be good to do a short design proposal.

**Comment by nitgoy** [comment]
> I would rather just build the WAF filter even if initially it just does one thing and we allow actions to be triggered by metadata from other filters. But I think it would be good to do a short design proposal.

Is there an approximate ETA for the WAF filter? Asking since I don't see any updates on the WAF issue/PR. The WAF design would also need to include triggers from filter metadata - I didn't see that in the proposed design/proto of that filter.

**Comment by mattklein123** [comment]
There is no ETA. It needs to get worked on.

**Comment by nitgoy** [comment]
I can create a simple WAF filter that does just this. Then someone can actually add the other functionalities of WAF outlined [here](https://github.com/envoyproxy/envoy/pull/11386). How does that sound? cc @kprakasam

**Comment by mattklein123** [comment]
Yup that would be great.

**Comment by garyyang2002** [comment]
Looks like there is no progress for WAF, is it ok to enable delay for rate limit filters?
